packages/kbn-object-versioning/lib/content_management_services_versioning.test.ts

```diff
@@ -0,0 +1,403 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { schema } from '@kbn/config-schema';
+import { set } from '@kbn/safer-lodash-set';
+import { get } from 'lodash';
+import { getTransforms } from './content_management_services_versioning';
+import type { ServiceDefinitionVersioned } from './content_management_types';
+
+/**
+ * Wrap the key with [] if it is a key from an Array
+ * @param key The object key
+ * @param isArrayItem Flag to indicate if it is the key of an Array
+ */
+const renderKey = (key: string, isArrayItem: boolean): string => (isArrayItem ? `[${key}]` : key);
+
+const flattenObject = (
+  obj: Record<any, any>,
+  prefix: string[] = [],
+  isArrayItem = false
+): Record<any, any> =>
+  Object.keys(obj).reduce<Record<any, any>>((acc, k) => {
+    const nextValue = obj[k];
+
+    if (typeof nextValue === 'object' && nextValue !== null) {
+      const isNextValueArray = Array.isArray(nextValue);
+      const dotSuffix = isNextValueArray ? '' : '.';
+
+      if (Object.keys(nextValue).length > 0) {
+        return {
+          ...acc,
+          ...flattenObject(
+            nextValue,
+            [...prefix, `${renderKey(k, isArrayItem)}${dotSuffix}`],
+            isNextValueArray
+          ),
+        };
+      }
+    }
+
+    const fullPath = `${prefix.join('')}${renderKey(k, isArrayItem)}`;
+    acc[fullPath] = nextValue;
+
+    return acc;
+  }, {});
+
+/**
+ * Create an object and set a value at the specific path
+ *
+ * @param path The path where to create the object
+ * @param value The value to set at the path
+ * @returns An object with a value at the provided path
+ */
+const setObjectValue = (path: string, value: unknown) => {
+  const obj = {};
+  set(obj, path, value);
+  return obj;
+};
+
+const wrapVersion = (obj: object, version = 1): object => ({
```

Comment:
Not sure if it would work, just thinking out loud:

```suggestion
const wrapVersion = <K extends number, V extends object>(obj: V, version: K = 1): {[version: K]: V} => ({
```